include/array/array.h

```diff
@@ -441,7 +443,17 @@ class dim : protected interval<Min_, Extent_> {
   /** Get or set the distance in flat indices between neighboring elements
    * in this dim. */
   NDARRAY_INLINE NDARRAY_HOST_DEVICE index_t stride() const { return stride_; }
-  NDARRAY_INLINE NDARRAY_HOST_DEVICE void set_stride(index_t stride) { stride_ = stride; }
+  NDARRAY_INLINE NDARRAY_HOST_DEVICE void set_stride(index_t stride) {
+    if (internal::is_static(Stride_)) {
+      if (internal::is_resolved(stride)) {
```

Comment:
I think this should be:

```
if (internal::is_static(Stride_)) {
  assert(!internal::is_resolved(stride) || stride == Stride_);
} else {
  stride_ = stride;
}
```
i.e. really do nothing if `Stride_` is static. I think we should do the same thing for `set_min` and `set_extent` (though they don't need the check for `is_resolved`). This way, dims with static properties have basically no-op constructors/assignments.

Reply:
Oh sure, that's more concise. Heh I had it that way but changed it to be more verbose. I'll fix it.

Reply:
You know my usual preference for immutable data structures. I never liked that you can mutate an `interval`. What do you think about adding, e.g., `interval::with_extent(index_t extent)` that returns a new instance?

Reply:
Re: min and extent - I added the check but realized that `constexpr_index` already handles this.

Reply:
I don't see that. It checks it when the value is 'read', not 'written', just like stride used to also. The problem with that is that it catches errors later than they could be caught, so it's harder to debug. `set_min` and `set_extent` look just like `set_stride` did before this PR. I think we should change that.

Reply:
Does [this](https://github.com/dsharlet/array/blob/master/include/array/array.h#L148) not assert the value when it's written? You're constructing / assigning to a `constexpr_index`.

Re: catching errors later than when they should be caught - I buy that - this will make debugging easier. But I think it also duplicates the same code.

Reply:
Oh, you're right, the implementation is just very different and I forgot about it.